blockstack/lib/nameset/namedb.py

```diff
@@ -1120,7 +1122,57 @@ def get_num_names_in_namespace( self, namespace_id ):
         cur = self.db.cursor()
         return namedb_get_num_names_in_namespace( cur, namespace_id, self.lastblock )
     
+    def get_num_names_in_namespace_at_block( self, namespace_id, block_id ):
+        """
+        Get the number of names in a namespace
+        """
+        cur = self.db.cursor()
+        return namedb_get_num_names_in_namespace( cur, namespace_id, block_id )
     
+    def get_v2_upgrade_threshold_block( self ):
+        """
+        Get the block ID (int) of when the name registration signal threshold was reached.
+        Returns `None` if the threshold has not been reached.
+        """
+        return namedb_get_v2_upgrade_threshold_block(self.db)
+
+    def set_v2_upgrade_threshold_block( self, block_id ):
+        """
+        Set the block ID of when the name registration signal threshold is reached.
+        """
+        namedb_set_v2_upgrade_threshold_block(self.db, block_id)
+
+    def get_v2_import_block_reached( self ):
+        """
+        Get the block ID of when the import block threshold is reached.
+        Returns `None` if the threshold has not been reached.
+        """
+        namedb_get_v2_import_block_reached(self.db)
+
+    def set_v2_import_block_reached( self, block_id ):
+        """
+        Set the block ID of when the name block import threshold is reached.
+        """
+        namedb_set_v2_import_block_reached(self.db, block_id)
+
+    def perform_v2_upgrade_datafile_export( self ):
+        """
+        Export a datafile used for the v2 upgrade.
+        """
+        # TODO: is this called in a way where db locking isn't an issue?
+        export_file_path = os.path.join( self.working_dir, 'v2_migration_data.tar.bz2')
+
+        block_id = self.get_current_block()
+
+        # self.make_backups(block_id)
+
+        from ..fast_sync import fast_sync_snapshot
+        # TODO: the -1 is required otherwise the backup files to export cannot be found. what is going on here?
```

Comment:
Couldn't figure out why `fast_sync_snapshot` needs `block_id - 1`. Without it, it can't find the paths reported by the `db.get_backup_dirs`, even when I call `make_backups` right beforehand.

Reply:
This won't work in prod.  The reason it works for `block_id - 1` is because the `fast_sync_snapshot()` method uses the fourth argument (`block_id - 1`) to identify a named database backup in the `~/.blockstack-server/backups` directory.  The test framework makes a backup _at every block_, whereas prod only does so once every 144 blocks.

You will need to manually trigger a back-up once the miner activation threshold is met.  This is normally handled within the `StateEngine` class as part of block-processing, but you'll need to trigger it manually here.  Something like this should work:

```
# override backup frequency to force a backup _now_
old_backup_frequency = self.backup_frequency
self.backup_frequency = block_id
self.make_backups(block_id)
self.backup_frequency = old_backup_frequency

# now you can do fast_sync_snapshot() with block_id
```

Reply:
Done!